rskj-core/src/main/java/co/rsk/net/discovery/PeerExplorer.java

```diff
@@ -379,24 +386,46 @@ private void removeConnections(List<PeerDiscoveryRequest> expiredRequests) {
             Node node = req.getRelatedNode();
 
             if (node != null) {
-                this.establishedConnections.remove(node.getId());
-                this.distanceTable.removeNode(node);
+                removeConnection(node);
             }
         }
     }
 
+    private void removeConnection(Node node) {
+        this.establishedConnections.remove(node.getId());
+        this.distanceTable.removeNode(node);
+        this.knownHosts.remove(node.getAddressAsString());
+    }
+
     private void addConnection(PongPeerMessage message, String ip, int port) {
         Node senderNode = new Node(message.getNodeId().getID(), ip, port);
-        if (!StringUtils.equals(senderNode.getHexId(), this.localNode.getHexId())) {
-            OperationResult result = this.distanceTable.addNode(senderNode);
-
-            if (result.isSuccess()) {
-                NodeID senderId = senderNode.getId();
-                this.establishedConnections.put(senderId, senderNode);
-                logger.debug("New Peer found ip:[{}] port[{}]", ip, port);
-            } else {
-                this.challengeManager.startChallenge(result.getAffectedEntry().getNode(), senderNode, this);
-            }
+        boolean isLocalNode = StringUtils.equals(senderNode.getHexId(), this.localNode.getHexId());
+        if (isLocalNode) {
+            return;
+        }
+
+        if (!this.allowMultipleConnectionsPerHost) {
+            disconnectPeerIfDuplicatedByNodeId(senderNode, ip, port);
+            // add only if flag set to false
+            this.knownHosts.put(senderNode.getAddressAsString(), senderNode.getId());
+        }
+
+        OperationResult result = this.distanceTable.addNode(senderNode);
+        if (result.isSuccess()) {
+            this.establishedConnections.put(senderNode.getId(), senderNode);
+            logger.debug("New Peer found or id changed: ip[{}] port[{}] id [{}]", ip, port, senderNode.getId());
+        } else {
+            this.challengeManager.startChallenge(result.getAffectedEntry().getNode(), senderNode, this);
+        }
+    }
+
+    private void disconnectPeerIfDuplicatedByNodeId(Node senderNode, String ip, int port) {
+        NodeID oldNodeIdForHost = this.knownHosts.get(senderNode.getAddressAsString());
+        boolean existsWithDifferentId = oldNodeIdForHost != null && !oldNodeIdForHost.equals(senderNode.getId());
+        if (existsWithDifferentId) {
+            logger.debug("Disconnecting peer with old id: ip[{}] port[{}] id [{}]", ip, port, oldNodeIdForHost);
```

Comment:
just thinking.. probably we may log a `warning` or `info` here (instead of `debug`)?

Reply:
Good point, I think it would be a weird behavior so I will make it a `warn`.